url73e3057d61_empeeoh_BACnet4J_tgz-pJ8-com_serotonin_bacnet4j_test_OverflowTestJ8/src/com/serotonin/bacnet4j/npdu/ip/WaitingRoom.java

```diff
@@ -0,0 +1,276 @@
+import com.serotonin.bacnet4j.apdu.AckAPDU;
+import com.serotonin.bacnet4j.apdu.ComplexACK;
+import com.serotonin.bacnet4j.apdu.ConfirmedRequest;
+import com.serotonin.bacnet4j.apdu.Segmentable;
+import com.serotonin.bacnet4j.exception.BACnetException;
+import com.serotonin.bacnet4j.exception.BACnetRuntimeException;
+import com.serotonin.bacnet4j.exception.BACnetTimeoutException;
+import com.serotonin.bacnet4j.exception.SegmentedMessageAbortedException;
+import com.serotonin.bacnet4j.type.constructed.Address;
+
+public class WaitingRoom {
+    private final HashMap<Key, Member> waitHere = new HashMap<Key, Member>();
+    private byte nextInvokeId;
+
+    synchronized public Key enterClient(InetSocketAddress peer, Address linkService) {
+        Member member = new Member();
+        Key key;
+
+        // Loop until we find a key that is available.
+        int attempts = 256;
+        while (true) {
+            // We set the server value in the key to true so that it matches with the message from the server.
+            key = new Key(peer, linkService, nextInvokeId++, true);
+
+            synchronized (waitHere) {
+                if (waitHere.get(key) != null) {
+                    // Key collision. Try again unless we've tried too many times.
+                    if (--attempts > 0)
+                        continue;
+                    throw new BACnetRuntimeException("Cannot enter a client into the waiting room. key=" + key);
+                }
+
+                // Found a good id. Use it and exit.
+                waitHere.put(key, member);
+                break;
+            }
+        }
+
+        return key;
+    }
+
+    public Key enterServer(InetSocketAddress peer, Address linkService, byte id) {
+        // We set the server value in the key to false so that it matches with the message from the client.
+        Key key = new Key(peer, linkService, id, false);
+        Member member = new Member();
+
+        synchronized (waitHere) {
+            if (waitHere.get(key) != null)
+                throw new BACnetRuntimeException("Cannot enter a server into the waiting room. key=" + key);
+            waitHere.put(key, member);
+        }
+
+        return key;
+    }
+
+    public AckAPDU getAck(Key key, long timeout, boolean throwTimeout) throws BACnetException {
+        return (AckAPDU) getAPDU(key, timeout, throwTimeout);
+    }
+
+    public ConfirmedRequest getRequest(Key key, long timeout, boolean throwTimeout) throws BACnetException {
+        return (ConfirmedRequest) getAPDU(key, timeout, throwTimeout);
+    }
+
+    public Segmentable getSegmentable(Key key, long timeout, boolean throwTimeout) throws BACnetException {
+        APDU apdu = getAPDU(key, timeout, throwTimeout);
+        if (apdu instanceof Abort)
+            throw new SegmentedMessageAbortedException((Abort) apdu);
+        try {
+            return (Segmentable) apdu;
+        }
+        catch (ClassCastException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)